woocommerce/wc-admin
# Replace useFilters with wp's withFilters

Eslint `react-hooks/rules-of-hooks` prevents the way we were using [a custom `useFilters` component](https://github.com/woocommerce/woocommerce-admin/blob/04774f21de707197a89d598bb18c8acee837c373/packages/components/src/higher-order/use-filters/README.md#L18).

It turns out `useFilters` was created before the (dark) ages of hooks and is actually a very outdated fork of `import { withFilters } from '@wordpress/components'`. `withFilters` has changed along the way and its filters now manipulate components instead of arrays or objects passed in filters. See the two README's below to see the differences in usage:

https://github.com/woocommerce/woocommerce-admin/tree/main/packages/components/src/higher-order/use-filters

https://github.com/WordPress/gutenberg/tree/master/packages/components/src/higher-order/with-filters

`withFilters` from Gutenberg now passes React components to the filter function so that extra components can be added to the UI. This is useful, but not the way we were using `useFilters`, which passed an array or object of data.

Our usage of `useFilters` was to trigger refreshes when filters were added. Considering these filters are run on application bootstrap and each page navigation, it seemed that refreshes aren't needed unless extensions add pages dynamically. This seems unlikely, but I've created a dummy extension to test this.

### Detailed test instructions:

Download the test extension and make sure the filters are working as they should:

[penguins.zip](https://github.com/woocommerce/woocommerce-admin/files/5083949/penguins.zip)

1. See the added page under WooCommerce called "penguins"
2. See the added report under Analytics called "penguins"
3. And see the Analytics setting added under Analytics > Settings. Note settings manipulation causes an error, but that is outside scope of this PR

### Changelog Note:

- dev: deprecate usage of component `useFilters` in favor of using wp.component's `withFilters`.
